### PR TITLE
Query History: Split data and view models

### LINKED ISSUE #44967: Query History: Split query history data models
Rich History uses `RichHistoryQuery` type as a view model (e.g. in `RichHistoryCard`) and the same object is saved to local storage. 

We're introducing backend storage for query history which will use a different data model structure. It will require changing a bit structure of objects used in the view layer:

| before | after |
|--------|------|
| `ts: number` - timestamp is used as unique id and date of creation | separate properties`createdAt: number` and `id: string`; for localstorage they both will be the same (just different type) but for backend storage they will be different|
| data source identified by `name` (which is unique across data sources but may change) | data source identified by `uid` which is unique and never changes |

At the same time we don't want to change the structure of existing items in local storage: there might be not enough space to change the structure. This means we need to split objects used in view layer and local storage.

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

For more details check #44967

This PR moves all properties from previous `RichHistoryQuery` into a local storage specific DTO object `RichHistoryLocalStorageDTO`. New `RichHistoryQuery` is used as a view model that is now used in view components. This way view layer doesn't know anything about how rich history is saved, it simply expects `RichHistoryQuery` which is converted in persistence layer and saved in storage exactly how it's been before.

When we add backend storage we will not need to change view components. We will add a new storage provider that will return `RichHistoryQuery` view models keeping storage specific details hidden from the view layer.

before

after

**Which issue(s) this PR fixes**:

Fixes #44967

**Special notes for your reviewer**:

No changes in the behavior. To test it check if the current history loads fine, add/edit/star/remove history items and play with filtering.

### CODE REVIEW

**Review by Elfo404 — CHANGES_REQUESTED** [review]

File: public/app/features/explore/RichHistory/RichHistoryCard.test.tsx [hunk 1]
```diff
@@ -32,15 +36,17 @@ const setup = (propOverrides?: Partial<Props>) => {
...
   datasourceName: 'Test datasource',
   starred: true,
   comment: 'test comment',
   queries: [
-    { query: 'query1', refId: 'A' },
-    { query: 'query2', refId: 'B' },
-    { query: 'query3', refId: 'C' },
+    { query: 'query1', refId: 'A' } as DataQuery,
+    { query: 'query2', refId: 'B' } as DataQuery,
+    { query: 'query3', refId: 'C' } as DataQuery,
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/RichHistory/RichHistoryCard.test.tsx]: Fixed in 6abee55ed3e63df3100086cc6acd4e6127df4ce4

File: public/app/features/explore/state/history.ts [hunk 2]
```diff
@@ -23,18 +23,26 @@ export const historyUpdatedAction = createAction<HistoryUpdatedPayload>('explore
...
-      nextRichHistory = await updateStarredInRichHistory(getState().explore.richHistory, ts);
+      nextRichHistory = await updateStarredInRichHistory(
+        getState().explore.richHistory,
+        id,
+        updatedProperty as boolean
+      );
     }
     if (property === 'comment') {
-      nextRichHistory = await updateCommentInRichHistory(getState().explore.richHistory, ts, updatedProperty);
+      nextRichHistory = await updateCommentInRichHistory(getState().explore.richHistory, id, updatedProperty as string);
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/state/history.ts]: Ended up splitting the logic into three separate functions in 497fc1698c4be1a89891143762a4a51ab7b26ecb

[the remark this answers is not in the document]
Comment (Elfo404) [reply public/app/features/explore/state/history.ts]: way cleaner implementation imho! 👍

**Review by Elfo404 — APPROVED** [review]
looks good! everything seems to work as before! 🚀

**Pull request merged**